exe/Main.hs

```diff
@@ -190,22 +184,404 @@ showEvent lock (EventFileDiagnostics (toNormalizedFilePath' -> file) diags) =
+    let consultCradle :: Maybe FilePath -> FilePath -> IO (IdeResult HscEnvEq)
+        consultCradle hieYaml cfp = do
+           logInfo logger $ T.pack ("Consulting the cradle for " <> show cfp)
+           cradle <- maybe (loadImplicitCradle $ addTrailingPathSeparator dir) loadCradle hieYaml
+           eopts <- cradleToSessionOpts cradle cfp
+           logDebug logger $ T.pack ("Session loading result: " <> show eopts)
+           case eopts of
+             -- The cradle gave us some options so get to work turning them
+             -- into and HscEnv.
+             Right opts -> do
+               res <- session (hieYaml, toNormalizedFilePath' cfp, opts)
+               return res
+             -- Failure case, either a cradle error or the none cradle
+             Left err -> do
+               dep_info <- getDependencyInfo ([fp | Just fp <- [hieYaml]])
+               let ncfp = toNormalizedFilePath' cfp
+               let res = (map (renderCradleError ncfp) err, Nothing)
+               modifyVar_ fileToFlags $ \var -> do
+                 pure $ Map.insertWith HM.union hieYaml (HM.singleton ncfp (res, dep_info)) var
+               return res
+
+    -- This caches the mapping from hie.yaml + Mod.hs -> [String]
+    let sessionOpts :: (Maybe FilePath, FilePath) -> IO (IdeResult HscEnvEq)
+        sessionOpts (hieYaml, file) = do
+          v <- fromMaybe HM.empty . Map.lookup hieYaml <$> readVar fileToFlags
+          cfp <- liftIO $ canonicalizePath file
+          case HM.lookup (toNormalizedFilePath' cfp) v of
+            Just (opts, old_di) -> do
+              deps_ok <- checkDependencyInfo old_di
+              if not deps_ok
+                then do
+                  -- If the dependencies are out of date then clear both caches and start
+                  -- again.
+                  modifyVar_ fileToFlags (const (return Map.empty))
+                  -- Keep the same name cache
+                  modifyVar_ hscEnvs (return . Map.adjust (\(h, _) -> (h, [])) hieYaml )
+                  consultCradle hieYaml cfp
+                else return opts
+            Nothing -> consultCradle hieYaml cfp
+
+    dummyAs <- async $ return (error "Uninitialised")
+    runningCradle <- newIORef dummyAs :: IO (IORef (Async (IdeResult HscEnvEq)))
+    -- The main function which gets options for a file. We only want one of these running
+    -- at a time. Therefore the IORef contains the currently running cradle, if we try
+    -- to get some more options then we wait for the currently running action to finish
+    -- before attempting to do so.
+    let getOptions :: FilePath -> IO (IdeResult HscEnvEq)
+        getOptions file = do
+            hieYaml <- cradleLoc file
+            sessionOpts (hieYaml, file)
+    -- The lock is on the `runningCradle` resource
+    lock <- newLock
+    return $ \file -> do
+      liftIO $ withLock lock $ do
+        as <- readIORef runningCradle
+        -- Is a cradle running?
+        finished <- poll as
```

Comment:
why do we first poll instead of just calling `wait` directly?

Reply:
Seems like that should work.